ARROW-11447: [Rust] Add shift kernel for primitive types

Dandandan opened the pull request:
This adds a `shift` function (now only accept primitive arrays) that shifts an array left or right by `offset`.

This is something that could be used to implement `LAG` `LEAD` etc. in DataFusion and other crates.

ritchie46 commented:
LGTM

Dandandan commented:
Hm, interesting indeed @ritchie46 I wrongly assumed that was pandas approach, but I now realize I recently saw `shift(1)` being used for implementing something similar to `LAG`. We should do whatever is standard I would say, maybe we should look at some other libraries as well.

nevi-me commented on rust/arrow/src/compute/kernels/window.rs:
```diff
@@ -0,0 +1,100 @@
+    T: ArrowPrimitiveType,
+{
+    // Compute slice
+    let slice_offset = clamp(offset, 0, values.len() as i64) as usize;
```
This is very neat :)

nevi-me reviewed (approved):
LGTM

@jorgecarleitao @alamb may you please take a look too

alamb commented on rust/arrow/src/compute/kernels/window.rs:
```diff
@@ -0,0 +1,100 @@
+    buffer::MutableBuffer,
+};
+
+/// Shifts array by defined number of items (to left or right)
```
I wonder if you can add a doc example here, something like https://github.com/apache/arrow/blob/master/rust/arrow/src/compute/kernels/take.rs#L62-L76 just to show the behavior (lifting / simplifying one of unit tests is probably a good level)

alamb replied:
It might also be good to specify what happens if you shift by more than the size of the array (e.g. that there is no error, just the entire array is returned as null)

alamb commented on rust/arrow/src/compute/kernels/window.rs:
```diff
@@ -0,0 +1,100 @@
+
+        let b = res.as_any().downcast_ref::<Int32Array>().unwrap();
+        assert_eq!(a.len(), res.len());
+        assert_eq!(false, b.is_valid(0));
+        assert_eq!(1, b.value(1));
+        assert_eq!(false, b.is_valid(2));
```
```suggestion
        let expected: Int32Array = vec![None, Some(1), None].into();

        assert_eq!(b, &expected);
```

alamb replied:
FWIW I agree with @ritchie46  that I expect `shift([1,NULL,4], -1)` to shift LEFT -- so `[NULL, 4, NULL]` rather than right ( aka [NULL, 1, NULL]`)

I don't think Postgres has any similar function I could find to guide us. I would vote to be consistent with Pandas -- which appears to shift right with a positive shift: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shift.html

alamb commented on rust/arrow/src/compute/kernels/window.rs:
```diff
@@ -0,0 +1,100 @@
+        let res = shift(&a, 1).unwrap();
+
+        let b = res.as_any().downcast_ref::<Int32Array>().unwrap();
+        assert_eq!(a.len(), res.len());
+        assert_eq!(false, b.is_valid(0));
+        assert_eq!(4, b.value(1));
+        assert_eq!(false, b.is_valid(2));
```
```suggestion
        let expected: Int32Array = vec![None, Some(1), None].into();

        assert_eq!(b, &expected);
```

alamb reviewed (commented):
Thanks @Dandandan  -- I think this is a great addition. I think the sign direction of `shift` should be reversed (positive shifts to the right), though I would be happy to change my opinion if others felt differently or if there was some other tool / precedent for this sign convention

Dandandan commented:
I agree it is best to have the order the same as pandas, it also seems other libraries do LAG / right shift for positive values as well.

Cleaned it up a bit, added a doc test.

alamb commented on rust/arrow/src/compute/kernels/window.rs:
```diff
@@ -0,0 +1,110 @@
+/// use arrow::compute::shift;
+///
+/// let a: Int32Array = vec![Some(1), None, Some(4)].into();
+/// // shift array 1 element to the left
```
```suggestion
/// // shift array 1 element to the right
```

alamb commented:
Looking good -- thanks @Dandandan

alamb commented:
Looks like there is an error in the simd tests: https://github.com/apache/arrow/pull/9381/checks?check_run_id=1838559432

```


---- src/compute/kernels/window.rs - compute::kernels::window::shift (line 38) stdout ----
error[E0615]: attempted to take value of method `as_ref` on type `Arc<dyn Array>`
  --> src/compute/kernels/window.rs:47:16
   |
12 | assert_eq!(res.as_ref, &expected)
   |                ^^^^^^ method, not a field
   |
help: use parentheses to call the method
   |
12 | assert_eq!(res.as_ref(), &expected)
   |                      ^^

```

alamb commented:
I think this one is ready to go -- I'll plan to merge it tomorrow unless someone else beats me to it

alamb commented:
Thanks again @Dandandan

jorgecarleitao commented on rust/arrow/src/compute/kernels/window.rs:
```diff
@@ -0,0 +1,110 @@
+        T::DATA_TYPE,
+        nulls as usize,
+        Some(nulls),
+        Some(null_array.into()),
```
I did not have time to go through this, but I get the feeling that this buffer may be too large: imo it should be `ceil(nulls, 8)`, since there are 8 slots per byte for bitmaps, no?
